controllers/vmi_eviction_controller.go

```diff
@@ -0,0 +1,263 @@
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kubedrain "k8s.io/kubectl/pkg/drain"
+	kubevirtv1 "kubevirt.io/api/core/v1"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	"sigs.k8s.io/cluster-api/controllers/noderefutil"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	infrav1 "sigs.k8s.io/cluster-api-provider-kubevirt/api/v1alpha1"
+	context "sigs.k8s.io/cluster-api-provider-kubevirt/pkg/context"
+	"sigs.k8s.io/cluster-api-provider-kubevirt/pkg/workloadcluster"
+)
+
+const (
+	vmiDeleteGraceTimeoutDurationSeconds = 600 // 10 minutes
+)
+
+type VmiEvictionReconciler struct {
+	client.Client
+	workloadCluster workloadcluster.WorkloadCluster
+}
+
+// NewVmiEvictionReconciler creates a new VmiEvictionReconciler
+func NewVmiEvictionReconciler(cl client.Client) *VmiEvictionReconciler {
+	return &VmiEvictionReconciler{Client: cl, workloadCluster: workloadcluster.New(cl)}
+}
+
+// SetupWithManager will add watches for this controller.
+func (r *VmiEvictionReconciler) SetupWithManager(ctx goContext.Context, mgr ctrl.Manager) error {
+	selector, err := getLabelPredicate()
+
+	if err != nil {
+		return fmt.Errorf("can't setup the VMI eviction controller; %w", err)
+	}
+
+	_, err = ctrl.NewControllerManagedBy(mgr).
+		For(&kubevirtv1.VirtualMachineInstance{}).
+		WithEventFilter(selector).
+		Build(r)
+
+	return err
+}
+
+// +kubebuilder:rbac:groups=cluster.x-k8s.io,resources=clusters;machines,verbs=get;list;watch
+// +kubebuilder:rbac:groups="",resources=secrets;,verbs=get;list;watch
+// +kubebuilder:rbac:groups=kubevirt.io,resources=virtualmachineinstances;,verbs=get;list;watch;patch;update;delete
+
+// Reconcile handles VMI events.
+func (r VmiEvictionReconciler) Reconcile(ctx goContext.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := ctrl.LoggerFrom(ctx)
+
+	vmi := &kubevirtv1.VirtualMachineInstance{}
+	err := r.Get(ctx, req.NamespacedName, vmi)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			logger.V(4).Info(fmt.Sprintf("Can't find virtualMachineInstance %s; it was already deleted.", req.NamespacedName))
+			return ctrl.Result{}, nil
+		}
+		logger.Error(err, fmt.Sprintf("failed to read VMI %s", req.Name))
+		return ctrl.Result{}, err
+	}
+
+	if !shouldGracefulDeleteVMI(vmi, logger, req.NamespacedName) {
+		return ctrl.Result{}, nil
+	}
+
+	cluster, err := r.getCluster(ctx, vmi)
+	if err != nil {
+		logger.Error(err, "Can't get the cluster form the VirtualMachineInstance", "VirtualMachineInstance name", req.NamespacedName)
+		return ctrl.Result{}, err
+	}
+
+	nodeDrained, retryDuration, err := r.drainNode(ctx, cluster, vmi.Status.EvacuationNodeName, logger)
+	if err != nil {
+		return ctrl.Result{RequeueAfter: retryDuration}, err
+	}
+
+	if !nodeDrained && r.waitingForTimeout(ctx, vmi, logger) {
+		return ctrl.Result{RequeueAfter: retryDuration}, nil
+	}
```

Comment:
The drain can block indefinitely due to the way this logic is structured. The timeout needs to be checked either regardless if the drainNode() function returns an error or before drainNode is even called. If the global timeout occurs, the VMI should be deleted.

We don't want anything within the guest cluster to indefinitely block an infra cluster node from being drained.